### PR TITLE
Optimize the Importing of a world

### PR DESCRIPTION
**Pull request by 0xnim**

### CODE REVIEW

File: src/lib/storage/src/lmdb.rs [hunk 1]
```diff
@@ -159,6 +159,45 @@ impl LmdbBackend {
...
+            let keymap: HashMap<u128, &Vec<u8>> = data.iter().map(|(k, v)| (*k, v)).collect();
+
+            // Iterate through the keys in sorted order
+            let mut sorted_keys: Vec<u128> = keymap.keys().cloned().collect();
+            sorted_keys.sort();
+
+            // Iterate through the sorted keys to perform upserts
+            for key in sorted_keys {
+                // Check if the key already exists
+                if db.get(&rw_txn, &key)?.is_some() {
```
Comment (ReCore-sys) [inline src/lib/storage/src/lmdb.rs:184]: Why check but do the same thing anyway? That's just an extra IO operation

**Review by radstevee — APPROVED** [review]

File: src/lib/world/src/importing.rs [hunk 2]
```diff
@@ -1,172 +1,221 @@
...
-        let progress_bar = Arc::new(ProgressBar::new(self.get_chunk_count(import_dir)?));
-        info!("Importing chunks from import directory...");
-        let start = std::time::Instant::now();
-        let mut task_set = JoinSet::new();
+        check_paths_validity(&import_dir)?;
+
+        let total_chunks = self.get_chunk_count(&import_dir)?;
+        let progress_style = ProgressStyle::default_bar()
+            .template("[{elapsed_precise}] {bar:40.cyan/blue} {pos:>7}/{len:7} {msg}")
+            .unwrap();
```
Comment (ReCore-sys) [inline src/lib/world/src/importing.rs:82]: don't use unwrap

File: src/lib/world/src/db_functions.rs [hunk 3]
```diff
@@ -129,6 +130,31 @@ pub(crate) async fn save_chunk_internal(world: &World, chunk: Chunk) -> Result<(
...
 }
 
+pub(crate) async fn save_chunk_internal_batch(
+    world: &World,
+    chunks: Vec<Chunk>,
+) -> Result<(), WorldError> {
+    // Prepare the batch data for the upsert
+    let mut batch_data = Vec::new();
+
+    for chunk in chunks.iter() {
```
Comment (ReCore-sys) [inline src/lib/world/src/db_functions.rs:140]: Could do this as an iterator with rayon to multithread the compression

File: src/lib/world/src/importing.rs [hunk 4]
```diff
@@ -1,172 +1,221 @@
...
-    fn get_chunk_count(&self, import_dir: PathBuf) -> Result<u64, WorldError> {
+    fn get_chunk_count(&self, import_dir: &Path) -> Result<u64, WorldError> {
         info!("Counting chunks in import directory...");
         let regions_dir = import_dir.join("region").read_dir()?;
         let chunk_count = AtomicU64::new(0);
+
         regions_dir
-            .into_iter()
             .par_bridge()
-            .for_each(|region_file| match region_file {
```
Comment (ReCore-sys) [inline src/lib/world/src/importing.rs:57]: if you are removing these checks, make sure it still works if there isn't valid data to import and that there are clear messages if the data isn't valid

**Review by ReCore-sys — CHANGES_REQUESTED** [review]
Logic is sound, could do with some performance tune ups

File: src/lib/storage/src/lmdb.rs (hunk 1 above)
Comment (0xnim) [reply src/lib/storage/src/lmdb.rs:184]: because if you dont then lmdb has to do it anyway and lmdb does it slower and per each into the whole database

Comment (ReCore-sys) [reply src/lib/storage/src/lmdb.rs:184]: But you are doing the exact same function call, regardless of whether the condition fails or passes, so calling `get()` first won't make a difference.

Comment (0xnim) [reply src/lib/storage/src/lmdb.rs:184]: I tested

File: src/lib/world/src/importing.rs (hunk 2 above)
Comment (0xnim) [reply src/lib/world/src/importing.rs:82]: i have to? don't i?

Comment (ReCore-sys) [reply src/lib/world/src/importing.rs:82]: no? Print out the error with a description or return it if the function returns a result

File: src/lib/world/src/importing.rs (hunk 4 above)
Comment (0xnim) [reply src/lib/world/src/importing.rs:57]: should not be a problem

File: src/lib/world/src/importing.rs (hunk 2 above)
Comment (0xnim) [reply src/lib/world/src/importing.rs:82]: this is the progress  bar dude

**Review by ReCore-sys — APPROVED** [review]

**Comment by 0xnim** [comment]
ty

**Pull request merged**